File: workloads/Python/bst/impl.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple, List


@dataclass(eq=True)
class T:
    k: int
    v: int
    left: "BST" | None = None
    right: "BST" | None = None

    def __repr__(self) -> str:  # stable, Rust-like
        return f"(T {self.left} {self.k} {self.v} {self.right})"


@dataclass(eq=True)
class E:
    def __repr__(self) -> str:
        return "(E)"


BST = T | E


FUEL = 10_000


def _node(left: BST | None, k: int, v: int, right: BST | None) -> T:
    return T(k, v, left or E(), right or E())


# Insert
def insert(k: int, v: int, t: BST) -> BST:
    match t:
        case E():
            return _node(E(), k, v, E())
        case T(k2, v2, l, r):
            """! insert """
            if k < k2:
                return _node(insert(k, v, l), k2, v2, r)
            elif k2 < k:
                return _node(l, k2, v2, insert(k, v, r))
            else:
                return _node(l, k2, v, r)
            """!! insert_1 """
            """!
            return _node(E(), k, v, E())
            """
            """!! insert_2 """
            """!
            if k < k2:
                return _node(insert(k, v, l), k2, v2, r)
            else:
                return _node(l, k2, v, r)
            """
            """!! insert_3 """
            """!
            if k < k2:
                return _node(insert(k, v, l), k2, v2, r)
            elif k2 < k:
                return _node(l, k2, v2, insert(k, v, r))
            else:
                return _node(l, k2, v2, r)
            """
            """ !"""


# Join
def join(l: BST, r: BST) -> BST:
    match (l, r):
        case (E(), _):
            return r
        case (_, E()):
            return l
        case (T(k1, v1, l1, r1), T(k2, v2, l2, r2)):
            return _node(l1, k1, v1, _node(join(r1, l2), k2, v2, r2))


# Delete
def delete(k: int, t: BST) -> BST:
    match t:
        case E():
            return E()
        case T(k2, v2, l, r):
            """! delete """
            if k < k2:
                return _node(delete(k, l), k2, v2, r)
            elif k2 < k:
                return _node(l, k2, v2, delete(k, r))
            else:
                return join(l, r)
            """!! delete_4 """
            """!
            _ = v2
            if k < k2:
                return delete(k, l)
            elif k2 < k:
                return delete(k, r)
            else:
                return join(l, r)
            """
            """!! delete_5 """
            """!
            if k2 < k:
                return _node(delete(k, l), k2, v2, r)
            elif k < k2:
                return _node(l, k2, v2, delete(k, r))
            else:
                return join(l, r)
            """
            """ !"""


# Below
def below(k: int, t: BST) -> BST:
    match t:
        case E():
            return E()
        case T(k2, v2, l, r):
            if k <= k2:
                return below(k, l)
            else:
                return _node(l, k2, v2, below(k, r))


# Above
def above(k: int, t: BST) -> BST:
    match t:
        case E():
            return E()
        case T(k2, v2, l, r):
            if k2 <= k:
                return above(k, r)
            else:
                return _node(above(k, l), k2, v2, r)


# Union with fuel (mirrors Rust union_8 variant)
def union_(l: BST, r: BST, f: int) -> BST:
    if f == 0:
        return E()
    f1 = f - 1
    match (l, r):
        case (E(), r):
            return r
        case (l, E()):
            return l
        case (T(k1, v1, l1, r1), T(k2, v2, l2, r2)):
            """! union """
            if k1 == k2:
                return _node(union_(l1, l2, f1), k1, v1, union_(r1, r2, f1))
            elif k1 < k2:
                return _node(
                    union_(l1, below(k1, l2), f1),
                    k1,
                    v1,
                    union_(r1, _node(above(k1, l2), k2, v2, r2), f1),
                )
            else:
                return union_(_node(l2, k2, v2, r2), _node(l1, k1, v1, r1), f1)
            """!! union_6 """
            """!
            return _node(l1, k1, v1, _node(union_(r1, l2, f1), k2, v2, r2))
            """
            """!! union_7 """
            """!
            if k1 == k2:
                return _node(union_(l1, l2, f1), k1, v1, union_(r1, r2, f1))
            elif k1 < k2:
                return _node(l1, k1, v1, _node(union_(r1, l2, f1), k2, v2, r2))
            else:
                return union_(_node(l2, k2, v2, r2), _node(l1, k1, v1, r1), f1)
            """
            """!! union_8 """
            """!
            if k1 == k2:
                return _node(union_(l1, l2, f1), k1, v1, union_(r1, r2, f1))
            elif k1 < k2:
                return _node(
                    union_(l1, below(k1, l2), f1),
                    k1,
                    v1,
                    union_(r1, _node(above(k1, l2), k2, v2, r2), f1),
                )
            else:
                return union_(_node(l2, k2, v2, r2), _node(l1, k1, v1, r1), f1)
            """
            """ !"""


def union(l: BST, r: BST) -> BST:
    return union_(l, r, FUEL)


def find(k: int, t: BST) -> Optional[int]:
    match t:
        case E():
            return None
        case T(k2, v2, l, r):
            if k < k2:
                return find(k, l)
            elif k2 < k:
                return find(k, r)
            else:
                return v2


def size(t: BST) -> int:
    match t:
        case E():
            return 0
        case T(_, _, l, r):
            return 1 + size(l) + size(r)


# Helpers used by specs
def to_list(t: BST) -> List[Tuple[int, int]]:
    match t:
        case E():
            return []
        case T(k, v, l, r):
            return [*to_list(l), (k, v), *to_list(r)]

File: workloads/OCaml/BST/lib/impl.ml
type t = E | T of t * int * int * t

let rec insert (k : int) (v : int) (t : t) =
  match t with
  | E -> T (E, k, v, E)
  | T (l, k', v', r) ->
      (*! insert *)
(*!
      if k < k' then T (insert k v l, k', v', r)
      else if k' < k then T (l, k', v', insert k v r)
      else T (l, k', v, r)
*)
      (*!! insert_1 *)
      let _ = ignore (l, k', v', r, insert) in
      T (E, k, v, E)
      (*!! insert_2 *)
      (*!
      if k < k' then T ((insert k v l), k', v', r)
      else T (l, k', v, r)
      *)
      (*!! insert_3 *)
      (*!
      if k < k' then T ((insert k v l), k', v', r)
      else if k' < k then T (l, k', v', (insert k v r))
      else T (l, k', v', r)
      *)
      (* !*)

let rec join (l : t) (r : t) =
  match (l, r) with
  | E, _ -> r
  | _, E -> l
  | T (l, k, v, r), T (l', k', v', r') -> T (l, k, v, T (join r l', k', v', r'))

let rec delete (k : int) (t : t) =
  match t with
  | E -> E
  | T (l, k', v', r) ->
      (*! delete *)
      if k < k' then T (delete k l, k', v', r)
      else if k' < k then T (l, k', v', delete k r)
      else join l r
      (*!! delete_4 *)
      (*!
  let _ = ignore v' in
  if k < k' then delete k l
  else if k' < k then delete k r
  else join l r
      *)
      (*!! delete_5 *)
      (*!
  if k' < k then T ((delete k l), k', v', r)
  else if k < k' then T (l, k', v', (delete k r))
  else join l r
      *)
      (* !*)

let rec below (k : int) (t : t) =
  match (k, t) with
  | _, E -> E
  | k, T (l, k', v, r) -> if k <= k' then below k l else T (l, k', v, below k r)

let rec above (k : int) (t : t) =
  match (k, t) with
  | _, E -> E
  | k, T (l, k', v, r) -> if k' <= k then above k r else T (above k l, k', v, r)

let rec union (l : t) (r : t) =
  match (l, r) with
  | E, _ -> r
  | _, E -> l
  (*! union *)
  | T (l, k, v, r), t -> T (union l (below k t), k, v, union r (above k t))
  (*!! union_6 *)
  (*!
  | T (l, k, v, r), T (l', k', v', r') -> T (l, k, v, T (union r l', k', v', r'))
  *)
  (*!! union_7 *)
  (*!
  | T (l, k, v, r), T (l', k', v', r') ->
      if k == k' then T (union l l', k, v, union r r')
      else if k < k' then T (l, k, v, T (union r l', k', v', r'))
      else union (T (l', k', v', r')) (T (l, k, v, r))
  *)
  (*!! union_8 *)
  (*!
  | T (l, k, v, r), T (l', k', v', r') ->
      if k == k' then T (union l l', k, v, union r r')
      else if k < k' then
        T (union l (below k l'), k, v, union r (T (above k l', k', v', r')))
      else union (T (l', k', v', r')) (T (l, k, v, r))
  *)
  (* !*)

let rec find (k : int) (t : t) : int option =
  match (k, t) with
  | _, E -> None
  | k, T (l, k', v', r) ->
      if k < k' then find k l else if k' < k then find k r else Some v'

let rec size (t : t) =
  match t with E -> 0 | T (l, _, _, r) -> 1 + size l + size r

(** Helper functions *)

let rec keys (t : t) : int list =
  match t with
  | E -> []
  | T (l, k, _v, r) ->
      let lk = keys l in
      let rk = keys r in
      [ k ] @ lk @ rk

let rec is_bst (t : t) : bool =
  let open List in
  match t with
  | E -> true
  | T (l, k, _, r) ->
      is_bst l && is_bst r
      && for_all (fun k' -> k' < k) (keys l)
      && for_all (fun k' -> k' > k) (keys r)

(** Removes from a key-value list *)
let delete_key k kvs = List.filter (fun (k', _) -> k <> k') kvs

(** Insert into sorted key-value list. Replaces value if the key exists *)
let rec l_insert k v kvs =
  match kvs with
  | [] -> [ (k, v) ]
  | (k', v') :: kvs' ->
      if k < k' then (k, v) :: kvs
      else if k = k' then (k, v) :: kvs'
      else (k', v') :: l_insert k v kvs'

let rec is_sorted l =
  match l with
  | [] -> true
  | (k, _) :: l' -> (
      match l' with [] -> true | (k', _) :: _l'' -> k < k' && is_sorted l')

let rec to_list t =
  match t with E -> [] | T (l, k, v, r) -> to_list l @ [ (k, v) ] @ to_list r

let rec l_union l1 l2 =
  match l1 with
  | [] -> l2
  | (k, v) :: l1' ->
      let l2' = List.filter (fun (k', _) -> k <> k') l2 in
      (k, v) :: l_union l1' l2'

let eq t1 t2 = to_list t1 = to_list t2
let ( === ) t1 t2 = eq t1 t2

(** Others *)
let to_string t =
  let rec aux t =
    match t with
    | E -> "(E)"
    | T (l, k, v, r) -> Printf.sprintf "(T %s %d %d %s)" (aux l) k v (aux r)
  in
  aux t
